models/repo_watch.go

```diff
@@ -74,66 +74,143 @@ func (repo *Repository) GetWatchers(page int) ([]*User, error) {
 	return users, sess.Find(&users)
 }
 
-func notifyWatchers(e Engine, act *Action) error {
-	// Add feeds for user self and all watchers.
-	watches, err := getWatchers(e, act.RepoID)
-	if err != nil {
-		return fmt.Errorf("get watchers: %v", err)
-	}
-
-	// Add feed for actioner.
-	act.UserID = act.ActUserID
-	if _, err = e.InsertOne(act); err != nil {
-		return fmt.Errorf("insert new actioner: %v", err)
-	}
-
-	act.loadRepo()
-	// check repo owner exist.
-	if err := act.Repo.getOwner(e); err != nil {
-		return fmt.Errorf("can't get repo owner: %v", err)
-	}
+func notifyWatchers(e Engine, actions ...*Action) error {
+	var watchers []*Watch
+	var repo *Repository
+	var err error
+	var permCode []bool
+	var permIssue []bool
+	var permPR []bool
+
+	for _, act := range actions {
+		repoChanged := repo == nil || repo.ID != act.RepoID
+
+		if repoChanged {
+			// Add feeds for user self and all watchers.
+			watchers, err = getWatchers(e, act.RepoID)
+			if err != nil {
+				return fmt.Errorf("get watchers: %v", err)
+			}
+		}
 
-	// Add feed for organization
-	if act.Repo.Owner.IsOrganization() && act.ActUserID != act.Repo.Owner.ID {
-		act.ID = 0
-		act.UserID = act.Repo.Owner.ID
+		// Add feed for actioner.
+		act.UserID = act.ActUserID
 		if _, err = e.InsertOne(act); err != nil {
 			return fmt.Errorf("insert new actioner: %v", err)
 		}
-	}
 
-	for i := range watches {
-		if act.ActUserID == watches[i].UserID {
-			continue
+		if repoChanged {
+			act.loadRepo()
+			repo = act.Repo
+
+			// check repo owner exist.
+			if err := act.Repo.getOwner(e); err != nil {
+				return fmt.Errorf("can't get repo owner: %v", err)
+			}
+		} else if act.Repo == nil {
+			act.Repo = repo
+		}
+
+		// Add feed for organization
+		if act.Repo.Owner.IsOrganization() && act.ActUserID != act.Repo.Owner.ID {
+			act.ID = 0
+			act.UserID = act.Repo.Owner.ID
+			if _, err = e.InsertOne(act); err != nil {
+				return fmt.Errorf("insert new actioner: %v", err)
+			}
 		}
 
-		act.ID = 0
-		act.UserID = watches[i].UserID
-		act.Repo.Units = nil
+		if len(actions) == 1 {
+			// More efficient to just check the code and not cache
+			for _, watcher := range watchers {
```

Comment:
I'm not very fond of handling this as a special case here. Looks quite messy; the `range` logic is broken. It's like a big `goto`. 😁

The way I see it, golang is very fast for moving around structures, slices, etc.. Querying the database, on the other hand, is the big bottleneck here. So it's very good that you've grouped the requests when there are many of them, but I don't see the point of handling the single element case separately.

Reply:
It was about not storing things in memory - not really about speed - but it's fine I'll remove it.